### DISCUSSION TITLE
how to set a request timeout? (not connectionTimeout)

### DISCUSSION THREAD

**Discussion by GioMakes** [Kong/unirest-java#519] [General]
Setting the connectionTimeout seems to be pretty obvious the timeout to connect a http client to the base URL.
But how to set how long a request may take?
In my case it seems that the request gets a TimeoutException after about 10s. 
Like for a connectionTimeout. Are there two parameter for http connecting and request duration?
Btw I am using v4.2.9

**Comment by ryber** [comment]
There is no such setting available that would timeout a active connection after some period. However you can use the async methods and do this with the future like this example that times out after 12 seconds

```java
   Unirest.get("http://localhost/something")
                .asStringAsync()
                .get(12, TimeUnit.SECONDS)
                .getBody();

```

**Comment by GioMakes** [comment]
Hi thx a lot, thats just what I have tried as workaround. No problem with that.
What I am concerned about is that I solved the timeout problem in long running REST calls (up to 25s) by rising the connectionTimeout to 30s.
These timeouts are for sure related in some way to the connectionTime while they are not originated to any kind of connection.
And while most of unirest api is highly intuitiv this connectionTimeout sound misleading to me.
But it runs smooth and I am happy :-)

Am 21.03.2024 um 21:14 schrieb Ryan Bergman ***@***.***>:



There is no such setting available that would timeout a active connection after some period. However you can use the async methods and do this with the future like this example that times out after 12 seconds

   Unirest.get("http://localhost/something")
                .asStringAsync()
                .get(12, TimeUnit.SECONDS)
                .getBody();

**Reply by ryber** [reply]
The setting is a direct proxy to the Java HTTPClient setting of the same name. 
The documentation for that reads:

```java
/**
         * Sets the connect timeout duration for this client.
         *
         * <p> In the case where a new connection needs to be established, if
         * the connection cannot be established within the given {@code
         * duration}, then {@link HttpClient#send(HttpRequest,BodyHandler)
         * HttpClient::send} throws an {@link HttpConnectTimeoutException}, or
         * {@link HttpClient#sendAsync(HttpRequest,BodyHandler)
         * HttpClient::sendAsync} completes exceptionally with an
         * {@code HttpConnectTimeoutException}. If a new connection does not
         * need to be established, for example if a connection can be reused
         * from a previous request, then this timeout duration has no effect.
         *
         * @param duration the duration to allow the underlying connection to be
         *                 established
         * @return this builder
         * @throws IllegalArgumentException if the duration is non-positive
         */
        public Builder connectTimeout(Duration duration);
```

**Comment by ryber** [comment]
So it turns out there is a Request Timeout we have available. Its on the request itself, there is also now a global config option for it in 4.4.0 see https://github.com/Kong/unirest-java/issues/526
